modules/controller/dc-controller.js

```diff
@@ -107,131 +107,62 @@ class DCController {
-    //             data_set_id,
-    //             ot_json_object_id,
-    //             buyer_node_id: dv_node_id,
-    //             buyer_erc_id: dv_erc725_identity,
-    //             seller_node_id: this.config.identity,
-    //             seller_erc_id: this.config.erc725Identity.toLowerCase(),
-    //             price: sellingData.price,
-    //             purchase_id: '',
-    //             status: 'COMPLETED',
-    //         });
-    //         message = 'Data purchase successfully finalized!';
-    //         status = 'COMPLETED';
-    //     }
-    //
-    //
-    //     const response = {
-    //         handler_id,
-    //         status,
-    //         wallet: this.config.node_wallet,
-    //         message,
-    //         price: sellingData.price,
-    //         seller_node_id: this.config.identity,
-    //         seller_erc_id: this.config.erc725Identity,
-    //     };
-    //
-    //     const dataPurchaseResponseObject = {
-    //         message: response,
-    //         messageSignature: Utilities.generateRsvSignature(
-    //             JSON.stringify(response),
-    //             this.web3,
-    //             this.config.node_private_key,
-    //         ),
-    //     };
-    //
-    //     await this.transport.sendDataPurchaseResponse(
-    //         dataPurchaseResponseObject,
-    //         dv_node_id,
-    //     );
-    // }
+    async handlePermissionedDataReadRequest(message) {
+        const {
+            data_set_id, dv_erc725_identity, ot_object_id, handler_id, nodeId,
+        } = message;
+
+        const privateDataPermissions = await Models.data_trades.findAll({
+            where: {
+                data_set_id,
+                ot_json_object_id: ot_object_id,
+                buyer_node_id: nodeId,
+                status: 'COMPLETED',
+            },
+        });
+        if (!privateDataPermissions || privateDataPermissions.length === 0) {
+            throw Error(`You don't have permission to view objectId:
+            ${ot_object_id} from dataset: ${data_set_id}`);
+        }
+
+        const replayMessage = {
+            wallet: this.config.node_wallet,
+            handler_id,
+        };
+        const promises = [];
+        privateDataPermissions.forEach((privateDataPermisssion) => {
+            promises.push(this.graphStorage.findDocumentsByImportIdAndOtObjectId(
+                data_set_id,
+                privateDataPermisssion.ot_json_object_id,
+            ));
+        });
+        const otObjects = await Promise.all(promises);
+        replayMessage.ot_objects = otObjects;
+
+        privateDataPermissions.forEach(async (privateDataPermisssion) => {
+            await Models.data_sellers.create({
+                data_set_id,
+                ot_json_object_id: privateDataPermisssion.ot_json_object_id,
+                seller_node_id: nodeId.toLowerCase(),
+                seller_erc_id: Utilities.normalizeHex(dv_erc725_identity),
```

Comment:
Move normalizeHex outside of the loop, and then use its value. So that we call this method only once